File: rust/src/leet_code/easy/merge_two_lists.rs
// rust/src/leet_code/easy/merge_two_lists.rs

// 21. Merge Two Sorted Lists
// https://leetcode.com/problems/merge-two-sorted-lists/

use crate::leet_code::common::list_node::ListNode;

#[derive(Debug)]
pub struct Solution;

impl Solution {
    #[allow(dead_code)]
    pub fn merge_two_lists(
        &self,
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        // let mut list1 = match (list1.as_ref(), list2.as_ref()) {// list1.as_ref() converts Option<Box<ListNode>> into Option<&ListNode>.
        //     (Some(node1), Some(node2)) => {
        //         if node1.val < node2.val {
        //             Some(Box::new((*node1).clone()))
        //         } else {
        //             Some(Box::new((*node2).clone()))
        //         }
        //     }
        //     (Some(node1), None) => Some(Box::new((*node1).clone())),
        //     (None, Some(node2)) => Some(Box::new((*node2).clone())),
        //     (None, None) => None,
        // };
        let mut list1 = list1;
        let mut list2 = list2;

        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;

        while list1.is_some() && list2.is_some() {
            // Safe to unwrap because we just checked is_some
            let v1 = list1.as_ref().unwrap().val;
            let v2 = list2.as_ref().unwrap().val;

            if v1 <= v2 {
                // Move list1 node into the merged list
                let mut node = list1.take().unwrap();
                list1 = node.next.take();
                tail.next = Some(node);
            } else {
                // Move list2 node into the merged list
                let mut node = list2.take().unwrap();
                list2 = node.next.take();
                tail.next = Some(node);
            }

            // Move tail to the last node
            tail = tail.next.as_mut().unwrap();
        }

        // Append the remaining part
        tail.next = if list1.is_some() { list1 } else { list2 };

        dummy.next
    }
}

#[cfg(test)]
mod tests {
    use crate::leet_code::common::list_node::{list_to_vec, vec_to_list};

    use super::*;

    #[test]
    fn test1() {
        let list1 = vec_to_list(vec![1, 2, 4]);
        let list2 = vec_to_list(vec![1, 3, 4]);
        let s = Solution;
        let merged = s.merge_two_lists(list1, list2);
        assert_eq!(list_to_vec(merged), vec![1, 1, 2, 3, 4, 4]);
    }

    #[test]
    fn test2() {
        let list1 = vec_to_list(vec![]);
        let list2 = vec_to_list(vec![]);
        let s = Solution;
        let merged = s.merge_two_lists(list1, list2);
        assert_eq!(list_to_vec(merged), vec![]);
    }

    #[test]
    fn test3() {
        let list1 = vec_to_list(vec![]);
        let list2 = vec_to_list(vec![0]);
        let s = Solution;
        let merged = s.merge_two_lists(list1, list2);
        assert_eq!(list_to_vec(merged), vec![0]);
    }
}

File: python/src/dsa/linked_list/merge_two_lists.py
# python/src/dsa/linked_list/merge_two_lists.py

# 21. Merge Two Sorted Lists
# https://leetcode.com/problems/merge-two-sorted-lists/description/

from typing import Optional

from dsa.linked_list.linked_list import (
    ListNode,
    linked_list_to_list,
    list_to_linked_list,
)


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        head = ListNode()
        current = head
        while list1 and list2:
            if list1.val < list2.val:
                current.next = ListNode(list1.val)
                list1 = list1.next
            else:
                current.next = ListNode(list2.val)
                list2 = list2.next

            current = current.next

        if list1:
            current.next = list1
        elif list2:
            current.next = list2

        return head.next

    def mergeTwoLists2(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        head = list1
        while list1 and list2:
            if list1.val < list2.val:
                list1 = list1.next
            else:
                # swap values
                list1_val = list1.val
                list1.val = list2.val
                list2.val = list1_val
                # insert list2 node to list1
                list2_next = list2.next
                list2.next = list1.next
                list1.next = list2
                list1 = list2
                list2 = list2_next

        if not head:
            return list2

        if list2:
            previous, current = head, head
            while current:
                previous = current
                current = current.next

            previous.next = list2

        return head

    def mergeTwoLists3(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        head = ListNode()
        current = head
        while list1 and list2:
            if list1.val < list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next

            current = current.next

        if list1:
            current.next = list1
        elif list2:
            current.next = list2

        return head.next


def main():
    list1 = list_to_linked_list([1, 2, 4])
    list2 = list_to_linked_list([1, 3, 4])
    print(linked_list_to_list(Solution().mergeTwoLists2(list1, list2)))


if __name__ == "__main__":
    main()
